zerver/views/auth.py

```diff
@@ -444,6 +447,21 @@ def oauth_redirect_to_root(request: HttpRequest, url: str,
 
 def start_social_login(request: HttpRequest, backend: str, extra_arg: Optional[str]=None
                        ) -> HttpResponse:
+    user_agent = parse_user_agent(request.META.get("HTTP_USER_AGENT", "Missing User-Agent"))
+    if user_agent["name"] == "ZulipElectron":
```

Comment:
I guess this check probably needs to have some sort of stronger detection than just ZulipElectron, since older versions won't support doing the decryption.  Maybe we can merge this with `>=5.1.0` here.

Do we want this to be purely a version check, or should we ask for the desktop app itself to do some sort of opt-in of support for the feature (i.e. that would allow us to disable it from a desktop app update, not just from a Zulip server update)?  I guess we could always disable it client-side by adding some hacky JS to mess up the login page in the desktop app side if we really needed to (or change the user-agent, but yuck).

Reply:
One solution is to do feature detection on `electron_bridge.decrypt_clipboard` inside `desktop-login.js`. If that’s not available, we can either

* show an error message telling the user to upgrade, or
* redirect back to `start_social_login` with something like `?force_in_app_login=1`, or
* present a form for the user to manually paste the token, and decrypt it with the Web Cryptography API.

Reply:
Yeah, I think it's reasonable to do that detection in `desktop_login.js`, which is code we expect to only run in the desktop app, rather than adding a hack in the login page (and adjust the parameters sent to this button).

Reply:
This now supports fallback to manual paste and Web Cryptography in the unmodified desktop app. (Still needs test fixes and other cleanup.)

Reply:
Great!